core/app/Utils.hs

```diff
@@ -27,3 +36,40 @@ defaultPrograms =
            , ghcPkgProgram = "ghc-pkg"
            , cabalProgram = "cabal"
            , stackProgram = "stack" }
+
+constructCells :: [String] -> Int -> [Cell] -> [Cell]
+constructCells [] _ curr = map (\c -> c { cellContent = T.strip $ cellContent c}) filtered
+  where filtered = filter (\c -> (T.stripStart $ cellContent c) /=  (T.pack "")) curr
+constructCells lst@(x:xs) i curr =
+  if isPrefixOf "--" x then
+    constructCells lst' (i+1) (curr ++ [textcell])
+  else
+    constructCells lst'' (i+1) (curr ++ [codecell])
+  where
+    lst' = dropWhile (\x -> isPrefixOf "--" x) lst
+    textcell = Cell TextCell i (T.pack $ unlines $ map (\('-':'-':' ':xs) -> xs) $ takeWhile (\x -> isPrefixOf "--" x) lst)
+    lst'' = dropWhile (\x -> not $ isPrefixOf "--" x) lst
+    codecell = Cell CodeCell i (T.pack $ unlines $ takeWhile (\x -> not $ isPrefixOf "--" x) lst)
+
```

Comment:
The `constructCells` looks **really** big, also the bindings aren't really meaningful. For instance, `lst'` should be something like `textCellList` and `lst''` should be `codeCellList`. `curr` -> `current`. We, functional programmers tend to run away from long binding names, but I think that it is not correct to do so when they are not specified by types.

For instance, the second argument, called `i` does not denotate anything, neither it's type. How can be this improved? Really easily:
- First, the most obvious way, is to change the name `i` to `cellInConstructionIndex`.
- Second, the less obvious way, but semantically better, is to switch from the type `Int` to a `newtype CellInConstructionIndex = CellInConstructionIndex Int`. This brings our code much better semantics, not only allowing us to call a function like `constructCells textCellList (CellInConstructionIndex (i+1)) (curr ++ [textCell])`

This brings another interesting idea to mind, *do we really need to recurse?* In my experience, generally, in FP it is better to avoid recursion, unless it's the most logical way to solve something. We are lucky that GHC implements [TCO](http://stackoverflow.com/questions/310974/what-is-tail-call-optimization), but if we switch to another language like Clojure or Scala, we might not be that lucky.

In this exact function, we are recursing to generate `CellId`s that go from 0 to the *total number of possible cells* in the list `lst`. But, whats this "possible cells" stuff?

Imagine this code, that is saved in a file:

```haskell
-- # Some markdown
-- ## This is an example
-- Foo
-- **Bar**
now :: This -> Is -> Some
code = 52

main :: IO
main = undefined
-- More text
-- And more
```

Instead of iterating over the lines and telling it **how** to do stuff, let's take a declarative approach, telling it **what** to do, in the end this is Haskell right? 😄 

So, what can we do? We need to **split** stuff, right? Let's take a look at the `Data.List.Split` module...
...it looks like it has [exactly what we need!](https://hackage.haskell.org/package/split-0.2.3.1/docs/Data-List-Split.html#v:divvy) (right under "Splitting combinators").

So something like `split (dropBlanks . condense $ whenElt (isPrefixOf "--")) lst` would give us a **list of strings** where the lines are grouped depending on their possible cell type. But wait, this line of code is kinda ugly too, let's refactor it a bit:

```haskell
listOfCellContentLines :: [String] -> [[String]]
listOfCellContentLines lst = split whenLineIsFromATextCell lst
  where
    whenLineIsFromATextCell = dropBlanks . condense $ whenElt (isPrefixOf "--")
```

If we wanted it to be perfect, we could even change the name `lst` to `fileLines`.

So, alright, we got rid of the reason to recurse due to line parsing, but what about indexing?! 😓 
Easy too!

```haskell
indexedCellContents :: [[String]] -> [(Int, [String])]
indexedCellContents cellContentLines = zip [1..length cellContents] cellContentLines
```

So right now what we need to build our cells, how do we do it? Well, the most reasonable way is to take this pair, which is ready to be "assembled" into a cell, and after a simple check, build a cell:

```haskell
buildCell :: (Int, [String]) -> Cell
buildCell (cellId, cellContentLines)
  | cellContentLines `areFrom` TextCell   = Cell TextCell cellId processedContents
  | cellContentLines `areFrom` CodeCell = Cell CodeCell cellId processedContents
  where
    areFrom TextCell = flip $ any (is PrefixOf "--")
    areFrom CodeCell = flip $ any (not . isPrefixOf "--")
    build TextCell = Cell TextCell cellId processedContents
    build CodeCell = Cell CodeCell cellId processedContents
    processedContents = T.pack . unlines . map (dropWhile commentCharacter)
    commentCharacter c = (c == '-') || (c == ' ')
```

Woah! A really big function!
Yes, and no. Actually the `buildCell` function is just *two lines*, **but** it has a huge `where`. That's okay sometimes. In other languages, like Java, one would declare themselves a `private` function inside of their class. Thanks that in Haskell we have "private function-scoped functions". The goal is to **optimize for reading and changing**. Remember that the code is always written from scratch once, but it can be read a dozen times or more.

So, how do we thread all of this now?

```haskell
constructCells :: [String] -> [Cell]
constructCells notebookLines = map buildCell . indexedCellContents $ listOfContentLines notebookLines
```

Look how simple it reads and **abstracts** over all the implementation details, now we can even *eta reduce* and prettify it a bit:

```haskell
constructCells :: [String] -> [Cell]
constructCells = map buildCell
               . indexedCellContents
               . listOfContentLines
```